distributed/diagnostics/status_monitor.py

```diff
@@ -0,0 +1,103 @@
+from ..utils import ignoring, is_kernel
+from ..executor import default_executor
+from ..scheduler import Scheduler
+
+from tornado.httpclient import AsyncHTTPClient
+from tornado.ioloop import IOLoop, PeriodicCallback
+
+with ignoring(ImportError):
+    from bokeh.models import ColumnDataSource
+    from bokeh.models.widgets import DataTable, TableColumn
+    from bokeh.plotting import vplot, output_notebook, show
+    from bokeh.io import curstate, push_notebook
+
+class Status_Monitor(object):
+    """ Display the tasks running and waiting on each worker
+
+    Parameters
+    ----------
+    addr: tuple, optional
+        (ip, port) of scheduler.  Defaults to scheduler of recent Executor
+    interval: Number, optional
+        Interval between updates.  Defaults to 1s
+    """
+    def __init__(self, addr=None, interval=1000.00):
+        if addr is None:
+            scheduler = default_executor().scheduler
+            if isinstance(scheduler, rpc):
+                addr = (scheduler.ip, scheduler.port) # doesn't work
+            elif isinstance(scheduler, Scheduler):
+                addr = ('127.0.0.1', scheduler.services['http'].port)
+
+        self.display_notebook = False
+
+        if is_kernel() and not curstate().notebook:
+            output_notebook()
+            assert curstate().notebook
+
+        wkr_props = ['workers', 'ncores', 'nbytes', 'processing']
+        self.wkr_source = ColumnDataSource({k: [] for k in wkr_props})
+
+        wkr_columns = [TableColumn(field=prop, title=prop) for prop in wkr_props]
+        self.worker_table = DataTable(source=self.wkr_source, columns=wkr_columns,
+                                      width=600, height=100)
+
+        task_props = ['waiting', 'ready', 'failed', 'in-progress',
+                      'in-memory', 'total']
+        self.task_source = ColumnDataSource({k: [] for k in task_props})
+
+        task_columns = [TableColumn(field=prop, title=prop) for prop in task_props]
+        self.task_table = DataTable(source=self.task_source, columns=task_columns,
+                                    row_headers=False, width=600, height=60)
+
+        self.client = AsyncHTTPClient()
+
+        loop = IOLoop.current()
+        cb = lambda: self.update(*addr)
+        loop.add_callback(cb)
+        pc = PeriodicCallback(cb, interval, io_loop=loop)
+        pc.start()
+
+    def _ipython_display_(self, **kwargs):
+        show(vplot(self.task_table, self.worker_table))
+        self.display_notebook = True
+
+    @gen.coroutine
+    def update(self, ip, port):
+        """ Query the Scheduler, update the figure
+
+        This opens a connection to the scheduler, sends it a function to run
+        periodically, streams the results back and uses those results to update
+        the bokeh figure
+        """
+
+        try:
+            response = yield self.client.fetch('http://%s:%d/status.json' % (ip, port))
+        except Exception as e:
+            print(e)
```

Comment:
In this case we probably want to return or raise (or just let the exception percolate up)